design/design-doc-composition-functions.md

```diff
@@ -209,570 +738,353 @@ metadata:
+  functions:
+  - name: compose-xr-using-go-templates
+    functionRef:
+      name: starlark
+    config:
+      apiVersion: example.org/v1
+      kind: StarlarkScript
+      script: ...
 ```
 
-A function that could not complete successfully could do so by returning the
-following `FunctionIO`:
+In future it may be useful for a Function to be able to deliver this type as a
+custom resource definition (CRD). This would allow a single configuration to be
+more easily shared by multiple Compositions. A Composition could reference a
+Function configuration custom resource:
 
 ```yaml
-apiVersion: apiextensions.crossplane.io/v1alpha1
-kind: FunctionIO
-config:
-  apiVersion: database.example.org/v1alpha1
-  kind: Config
-  metadata:
-    name: cloudsql
-  spec:
-    version: POSTGRES_9_6
-observed: {}  # Omitted for brevity.
-results:
-- severity: Error
-  message: "Could not render Database.postgresql.crossplane.io/v1beta1`
+  functions:
+  - name: compose-xr-using-go-templates
+    functionRef:
+      name: starlark
+    configRef:
+      apiVersion: example.org/v1
+      kind: StarlarkScript
+      name: make-some-purple-robots
 ```
 
-### Running Container Function Pipelines
+At this stage I suggest holding off on building this functionality until there
+is clear demand. 
 
-While Crossplane typically runs in a Kubernetes cluster - a cluster designed to
-run containers - running an ordered _pipeline_ of short-lived containers via
-Kubernetes is much less straightforward than you might expect. Refer to
-[Alternatives Considered](#alternatives-considered) for details.
+### Metrics and Tracing
 
-In order to provide flexibility and choice of tradeoffs in running containers
-(e.g. speed, scalability, security) this document proposes Crossplane defer
-containerized functions to an external runner. Communication with the runner
-would occur via a gRPC API, with the runner expected to be listening at the
-`endpoint` specified via the function's `runner` configuration block. This
-endpoint would default to `unix:///@crossplane/fn/default.sock` - an abstract
-[Unix domain socket][unix-domain-sockets].
+Crossplane does not currently expose its own metrics. It relies on [the set
+it gets from controller-runtime][controller-runtime-metrics], and metrics that
+may be [derived from the Kubernetes events it emits][event-exporter-metrics].
 
-Communication between Crossplane and a containerized function runner would use
-the following API:
+These metrics pertain to Kubernetes controllers, and will be insufficient when a
+significant portion of Composition logic is 'outsourced' to Functions. It will
+be important to establish a pattern for instrumenting Functions (e.g. as part of
+the Function SDKs). Request tracing in particular is likely to be useful in
+order to debug slow or failing Functions in a pipeline.
 
-```protobuf
-syntax = "proto3";
+Metrics and tracing for Functions must be implemented before the feature becomes
+generally available (GA).
```

Comment:
worth starting a GA vs Beta checklist in this doc (or tracking issue)?

Reply:
I'm inclined to use an issue for this since they're easier to update as we remember more things. 😄 

I think to start with if/when this is merged we'll want to update https://github.com/crossplane/crossplane/issues/3751 to reflect what needs to be done to hit beta. Maybe once we hit beta we can open another issue to track things we need to hit GA.